src/pubsub/mod_pubsub.erl

```diff
@@ -879,34 +879,63 @@ remove_user(Acc, User, Server) ->
+    [unsubscribe_pubsub_node_per_plugin_and_backend(Backend, Plugin, Entity, DbErrorCall, S) || S <- Subs].
+
+remove_user_affiliations_per_plugin_and_backend(Backend, Plugin, Host, Entity, DbErrorCall) ->
+    Affs = get_requested_data_per_plugin_and_backend(get_entity_affiliations, Backend, Plugin, Host, Entity, DbErrorCall),
+    [remove_affiliations_per_plugin_and_backend(Backend, Plugin, Entity, DbErrorCall, A) || A <- Affs].
+
+unsubscribe_pubsub_node_per_plugin_and_backend(Backend, Plugin, Entity, Err, {#pubsub_node{id = Nidx}, _, _, JID}) ->
+    ErrorDebug = Err#{action => unsubscribe_node,
+                      args => [Backend, Nidx, Entity, JID, all]
+     },
+    Backend:dirty(fun() -> Plugin:unsubscribe_node(Backend, Nidx, Entity, JID, all) end, ErrorDebug).
+
+get_requested_data_per_plugin_and_backend(Request, Backend, Plugin, Host, Entity, Err) ->
+    {result, Data} = Backend:dirty(fun() -> Plugin:Request(Backend, Host, Entity) end,
+                       Err#{action => Request, args => [Backend, Host, Entity]}),
+    Data.
+
+remove_affiliations_per_plugin_and_backend(Backend, Plugin, Entity, ErrorDebug, PubSubNode) ->
+    LServer = Entity#jid.lserver,
+    LUser = Entity#jid.luser,
     HomeTreeBase = <<"/home/", LServer/binary, "/", LUser/binary>>,
-    spawn(fun() -> lists:foreach(
-                     fun(PType) ->
-                             catch remove_user_per_plugin(PType, Host, Entity, HomeTreeBase)
-                     end, plugins(Host)) end).
-
-remove_user_per_plugin(PType, Host, Entity, HomeTreeBase) ->
-    {result, Subs} = node_action(Host, PType, get_entity_subscriptions, [Host, Entity]),
-    lists:foreach(fun({#pubsub_node{id = Nidx}, _, _, JID}) ->
-                          node_action(Host, PType, unsubscribe_node, [Nidx, Entity, JID, all]);
-                     (_) ->
-                          ok
-                  end, Subs),
-    {result, Affs} = node_action(Host, PType, get_entity_affiliations, [Host, Entity]),
-    lists:foreach(fun({#pubsub_node{nodeid = {H, N}, parents = []}, owner}) ->
-                          delete_node(H, N, Entity);
-                     ({#pubsub_node{nodeid = {H, N}, type = Type}, owner})
-                       when N == HomeTreeBase, Type == <<"hometree">> ->
-                          delete_node(H, N, Entity);
-                     ({#pubsub_node{id = Nidx}, publisher}) ->
-                          node_action(Host, PType,
-                                      set_affiliation,
-                                      [Nidx, Entity, none]);
-                     (_) ->
-                          ok
-                  end, Affs).
+    case should_delete_node(PubSubNode, HomeTreeBase) of
+        {true, {H, N}} -> delete_node(Backend, H, N, Entity);
+        {false, Nidx} -> Backend:dirty(fun() -> Plugin:set_affiliation(Backend, Nidx, Entity, none) end, ErrorDebug)
+    end.
+
+should_delete_node({#pubsub_node{id = Nidx, type = <<"flat">>}, owner}, _) ->
```

Comment:
good question, there should be test for that btw

Reply:
Ok, I just found the place where affiliations atoms are set, it's on the function `string_to_affiliation`, and the options are `owner`, `publisher`, `publish_only`, `member`, `outcast`, and `none`. And I verified the old implementation: only in the case of publisher it was resetting the affiliation to `none`. @fenek what should I do for all the other affiliations? I suppose that the removing logic is on the case of owner, and for all other affiliations we set it to `none`, is that correct? Or we ignore some affiliations in the logic?

Reply:
We don't remove any nodes in case of non-owner affiliations, so this function should simply return `false` for them. And now it will crash because it doesn't match for other affs than `owner` and `publisher`. I think that the last clause should be:

`should_delete_node({#pubsub_node{id = Nidx}, _}, _)`

Reply:
So I assume then that for all other affiliations that are not owner, we simply set this affiliation to none. I'm doing it that way then.